### PR TITLE
feat: AppProject synchronization

### PR DESCRIPTION
**Pull request by ishitasequeira**
Closes https://github.com/argoproj-labs/argocd-agent/issues/111

### CODE REVIEW

File: internal/backend/kubernetes/appproject/kubernetes.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+		}
+	} else {
+		l, err := be.appClient.ArgoprojV1alpha1().AppProjects("").List(ctx, v1.ListOptions{})
+		if err != nil {
+			return nil, err
+		}
+		res = append(res, l.Items...)
+	}
+	return res, nil
+}
```
Comment (jannfis) [inline internal/backend/kubernetes/appproject/kubernetes.go]: I think AppProjects will only ever live in Argo CD's installation namespaces, so there's no need to list them across namespaces here.

**Review by jannfis — COMMENTED** [review]
Starting an agent in managed mode from this code results in a panic, can you please check?

```
time="2024-09-10T14:21:41Z" level=info msg="Starting argocd-agent (agent) v99.9.9-unreleased (ns=agent-managed, allowed_namespaces=[], mode=managed)" module=Agent
time="2024-09-10T14:21:41Z" level=info msg="Outgoing unary call to /authapi.Authentication/Authenticate" module=Connector
time="2024-09-10T14:21:41Z" level=info msg="Starting app informer (namespaces: agent-managed)" module=AppInformer
time="2024-09-10T14:21:41Z" level=info msg="Starting app project informer (namespaces: )" module=AppProjectInformer
time="2024-09-10T14:21:41Z" level=info msg="App informer has shutdown" module=AppProjectInformer
time="2024-09-10T14:21:41Z" level=warning msg="Project Informer has exited" module=Agent
E0910 14:21:41.169145  138452 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 39 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x2fc00a0, 0x4bff940})
        /home/vagrant/go/pkg/mod/k8s.io/apimachinery@v0.29.6/pkg/util/runtime/runtime.go:75 +0x85
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x2540be400?})
        /home/vagrant/go/pkg/mod/k8s.io/apimachinery@v0.29.6/pkg/util/runtime/runtime.go:49 +0x6b
panic({0x2fc00a0?, 0x4bff940?})
        /home/vagrant/sdk/go1.22.3/src/runtime/panic.go:770 +0x132
k8s.io/client-go/tools/cache.(*Reflector).list(0xc000baa000, 0xc0009041e0)
        /home/vagrant/go/pkg/mod/k8s.io/client-go@v0.29.6/tools/cache/reflector.go:534 +0xadd
k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch(0xc000baa000, 0xc0009041e0)
        /home/vagrant/go/pkg/mod/k8s.io/client-go@v0.29.6/tools/cache/reflector.go:345 +0x28a
k8s.io/client-go/tools/cache.(*Reflector).Run.func1()
        /home/vagrant/go/pkg/mod/k8s.io/client-go@v0.29.6/tools/cache/reflector.go:291 +0x25
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x10?)
        /home/vagrant/go/pkg/mod/k8s.io/apimachinery@v0.29.6/pkg/util/wait/backoff.go:226 +0x33
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00090af50, {0x3851c40, 0xc000390fa0}, 0x1, 0xc0009041e0)
        /home/vagrant/go/pkg/mod/k8s.io/apimachinery@v0.29.6/pkg/util/wait/backoff.go:227 +0xaf
k8s.io/client-go/tools/cache.(*Reflector).Run(0xc000baa000, 0xc0009041e0)
        /home/vagrant/go/pkg/mod/k8s.io/client-go@v0.29.6/tools/cache/reflector.go:290 +0x1c5
k8s.io/client-go/tools/cache.(*controller).Run.(*Group).StartWithChannel.func2()
        /home/vagrant/go/pkg/mod/k8s.io/apimachinery@v0.29.6/pkg/util/wait/wait.go:55 +0x1b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
        /home/vagrant/go/pkg/mod/k8s.io/apimachinery@v0.29.6/pkg/util/wait/wait.go:72 +0x52
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 35
        /home/vagrant/go/pkg/mod/k8s.io/apimachinery@v0.29.6/pkg/util/wait/wait.go:70 +0x73
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
...
```

**Comment by ishitasequeira** [comment]
> Starting an agent in managed mode from this code results in a panic, can you please check?
> 
> ```
> time="2024-09-10T14:21:41Z" level=info msg="Starting argocd-agent (agent) v99.9.9-unreleased (ns=agent-managed, allowed_namespaces=[], mode=managed)" module=Agent
> time="2024-09-10T14:21:41Z" level=info msg="Outgoing unary call to /authapi.Authentication/Authenticate" module=Connector
> time="2024-09-10T14:21:41Z" level=info msg="Starting app informer (namespaces: agent-managed)" module=AppInformer
> time="2024-09-10T14:21:41Z" level=info msg="Starting app project informer (namespaces: )" module=AppProjectInformer
> time="2024-09-10T14:21:41Z" level=info msg="App informer has shutdown" module=AppProjectInformer
> time="2024-09-10T14:21:41Z" level=warning msg="Project Informer has exited" module=Agent
> E0910 14:21:41.169145  138452 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
> goroutine 39 [running]:
> k8s.io/apimachinery/pkg/util/runtime.logPanic({0x2fc00a0, 0x4bff940})
>         /home/vagrant/go/pkg/mod/k8s.io/apimachinery@v0.29.6/pkg/util/runtime/runtime.go:75 +0x85
> k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x2540be400?})
>         /home/vagrant/go/pkg/mod/k8s.io/apimachinery@v0.29.6/pkg/util/runtime/runtime.go:49 +0x6b
> panic({0x2fc00a0?, 0x4bff940?})
>         /home/vagrant/sdk/go1.22.3/src/runtime/panic.go:770 +0x132
> k8s.io/client-go/tools/cache.(*Reflector).list(0xc000baa000, 0xc0009041e0)
>         /home/vagrant/go/pkg/mod/k8s.io/client-go@v0.29.6/tools/cache/reflector.go:534 +0xadd
> k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch(0xc000baa000, 0xc0009041e0)
>         /home/vagrant/go/pkg/mod/k8s.io/client-go@v0.29.6/tools/cache/reflector.go:345 +0x28a
> k8s.io/client-go/tools/cache.(*Reflector).Run.func1()
>         /home/vagrant/go/pkg/mod/k8s.io/client-go@v0.29.6/tools/cache/reflector.go:291 +0x25
> k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x10?)
>         /home/vagrant/go/pkg/mod/k8s.io/apimachinery@v0.29.6/pkg/util/wait/backoff.go:226 +0x33
> k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00090af50, {0x3851c40, 0xc000390fa0}, 0x1, 0xc0009041e0)
>         /home/vagrant/go/pkg/mod/k8s.io/apimachinery@v0.29.6/pkg/util/wait/backoff.go:227 +0xaf
> k8s.io/client-go/tools/cache.(*Reflector).Run(0xc000baa000, 0xc0009041e0)
>         /home/vagrant/go/pkg/mod/k8s.io/client-go@v0.29.6/tools/cache/reflector.go:290 +0x1c5
> k8s.io/client-go/tools/cache.(*controller).Run.(*Group).StartWithChannel.func2()
>         /home/vagrant/go/pkg/mod/k8s.io/apimachinery@v0.29.6/pkg/util/wait/wait.go:55 +0x1b
> k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
>         /home/vagrant/go/pkg/mod/k8s.io/apimachinery@v0.29.6/pkg/util/wait/wait.go:72 +0x52
> created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 35
>         /home/vagrant/go/pkg/mod/k8s.io/apimachinery@v0.29.6/pkg/util/wait/wait.go:70 +0x73
> panic: runtime error: invalid memory address or nil pointer dereference [recovered]
>         panic: runtime error: invalid memory address or nil pointer dereference
> ...
> ```

Thanks @jannfis !! Surprisingly I did not see this as an issue on my workstation. Let me clean up the setup and restart from scratch to try and reproduce.

File: internal/informer/appproject/projectinformer.go [hunk 2]
```diff
@@ -179,5 +185,24 @@ func NewAppProjectInformer(ctx context.Context, client appclientset.Interface, o
...
+}
+
+func (i *AppProjectInformer) Start(ctx context.Context) {
+	log().Infof("Starting app project informer (namespaces: %s)", strings.Join(i.namespaces, ","))
+	err := i.projectInformer.Start(ctx)
+	if err != nil {
+		log().Errorf("Failed to start app project informer: %v", err)
+		return
+	}
+	log().Infof("project informer has shutdown")
```
Comment (jannfis) [inline internal/informer/appproject/projectinformer.go]: This log statement is misleading. The project informer keeps running in the background at this point in time.

File: internal/informer/informer.go [hunk 3]
```diff
@@ -285,8 +285,9 @@ func (i *GenericInformer) Start(ctx context.Context) error {
 		return fmt.Errorf("cannot start informer: already running")
 	}
 	i.setRunning(true)
-	i.logger.Debug("Starting informer goroutine")
+	i.logger.Info("Starting informer goroutine")
 	go i.informer.Run(i.runch)
+	i.logger.Info("Failed to start informer goroutine")
```
Comment (jannfis) [inline internal/informer/informer.go]: This log statement is misleading. The informer keeps running in the background.

**Review by jannfis — APPROVED** [review]
I'm gonna pull this in with only a superficial review to keep velocity high.

We can refine the code later on.

LGTM.

**Pull request merged**